#ifndef CARDCONTROL_H
#define CARDCONTROL_H

#pragma once

#include <QDialog>
#include <QTableWidget>
#include <QAbstractItemView>
#include <QStringList>
#include "cardoprotion.h"
#include "createstu.h"
#include "rechargedialog.h"
#include "inputstunumdialog.h"
#include "showstudialog.h"

namespace Ui {
class CardControl;
}

class CardControl : public QDialog
{
    Q_OBJECT

public:
    //===========================================================
    // 函数名：CardControl
    // 功能：CardControl类的构造函数，根据学生管理类中的学生信息更新
    //      窗口界面的组件stuTable，同时完成时间日期组件(dateTimeEdit)
    //      的初始化及其更新的设置
    // 输入参数：时间日期变量；父窗口部件的指针（没有父窗口则会传入nullptr）
    // 返回值：空
    //===========================================================
    explicit CardControl(QDateTime dt, QWidget *parent = nullptr);

    //===========================================================
    // 函数名：~CardControl
    // 功能：CardControl类的析构函数，在CardControl类的对象释放时释放ui
    // 输入参数：空
    // 返回值：空
    //===========================================================
    ~CardControl();

    //===========================================================
    // 函数名：updateTableWideget
    // 功能：更新该类的窗口界面的学生信息显示组件stuTable中所显示的信息
    // 输入参数：需要显示到表中的学生信息的键值对集合
    // 返回值：空
    //===========================================================
    void updateTableWideget(map<string,Student> students);

    //===========================================================
    // 函数名：matchWithNum
    // 功能：学生学号的模糊匹配，input为输入内容（其中*代表任意0-n个
    //      字符,?代表任意一个字符），stuNum为待比对的学号，进行input和
    //      stuNum的匹配并返回匹配结果
    // 输入参数：待比对学号；输入内容
    // 返回值：匹配成功与否
    //===========================================================
    bool matchWithNum(string stuNum, string input);

    //===========================================================
    // 函数名：matchWithName
    // 功能：学生姓名的模糊匹配，input为输入内容（其中?代表任意一个字符），
    //      name为待比对的姓名，进行input和stuNum的匹配并返回匹配结果
    // 输入参数：待比对姓名；输入内容
    // 返回值：匹配成功与否
    //===========================================================
    bool matchWithName(string name, string input);

private slots:

    //===========================================================
    // 函数名：on_goBack_clicked
    // 功能：该类窗口界面上的返回按钮的click()信号的槽函数，关闭窗口界面
    // 输入参数：空
    // 返回值：空
    //===========================================================
    void on_goBack_clicked();

    //===========================================================
    // 函数名：on_createStu_clicked
    // 功能：该类窗口界面上的开户按钮的click()信号的槽函数，显示开户窗口界面
    // 输入参数：空
    // 返回值：空
    //===========================================================
    void on_createStu_clicked();

    //===========================================================
    // 函数名：on_deleteStu_clicked
    // 功能：该类窗口界面上的开户按钮的click()信号的槽函数，显示销户窗口界面
    // 输入参数：空
    // 返回值：空
    //===========================================================
    void on_deleteStu_clicked();

    //===========================================================
    // 函数名：on_recharge_clicked
    // 功能：该类窗口界面上的充值按钮的click()信号的槽函数，显示充值窗口界面
    // 输入参数：空
    // 返回值：空
    //===========================================================
    void on_recharge_clicked();

    //===========================================================
    // 函数名：on_reportLoss_clicked
    // 功能：该类窗口界面上的挂失按钮的click()信号的槽函数，显示挂失窗口界面
    // 输入参数：空
    // 返回值：空
    //===========================================================
    void on_reportLoss_clicked();

    //===========================================================
    // 函数名：on_addCard_clicked
    // 功能：该类窗口界面上的补卡按钮的click()信号的槽函数，显示补卡窗口界面
    // 输入参数：空
    // 返回值：空
    //===========================================================
    void on_addCard_clicked();

    //===========================================================
    // 函数名：on_cancleFreeze_clicked
    // 功能：该类窗口界面上的解挂按钮的click()信号的槽函数，显示解挂窗口界面
    // 输入参数：空
    // 返回值：空
    //===========================================================
    void on_cancleFreeze_clicked();

    //===========================================================
    // 函数名：on_stuTable_itemDoubleClicked
    // 功能：该类窗口界面上的stuTable组件的单元格双击信号的槽函数，显示
    //      双击所在行的学生的信息显示界面
    // 输入参数：被双击的单元格的指针
    // 返回值：空
    //===========================================================
    void on_stuTable_itemDoubleClicked(QTableWidgetItem *item);

    //===========================================================
    // 函数名：timeUpdate
    // 功能：更新窗口界面中的dateTimeEdit组件的显示日期时间，每次更新加1ms
    // 输入参数：空
    // 返回值：空
    //===========================================================
    void timeUpdate();

    //===========================================================
    // 函数名：on_restoreBtn_clicked
    // 功能：恢复窗口组件stuTable上的显示信息为系统内所有学生的信息
    // 输入参数：空
    // 返回值：空
    //===========================================================
    void on_restoreBtn_clicked();

    //===========================================================
    // 函数名：on_stuNumCBox_stateChanged
    // 功能：根据学号勾选框的选择更新输入框、搜索按钮和姓名勾选框的设置
    // 输入参数：选择信号（2表示选择，0表示取消选择）
    // 返回值：空
    //===========================================================
    void on_stuNumCBox_stateChanged(int arg1);

    //===========================================================
    // 函数名：on_nameCBox_stateChanged
    // 功能：根据姓名勾选框的选择更新输入框、搜索按钮和学号勾选框的设置
    // 输入参数： 选择信号（2表示选择，0表示取消选择）
    // 返回值：空
    //===========================================================
    void on_nameCBox_stateChanged(int arg1);

    //===========================================================
    // 函数名：on_searchBtn_clicked
    // 功能：该类窗口界面上的搜索按钮的click()信号的槽函数，搜索出与lineEdit
    //      输入内容匹配的学生并将其显示在窗口组件stuTable上
    // 输入参数：空
    // 返回值：空
    //===========================================================
    void on_searchBtn_clicked();

private:
    Ui::CardControl *ui;
};

#endif // CARDCONTROL_H
